### ISSUE TITLE
Chromaticraft: Lost all progess after exiting nether.

### ISSUE THREAD

**Issue by FyeRee** [ReikaKalseki/Reika_Mods_Issues#1113]
While playing on my personal server I went ontop of the nether ceiling to look for some info fragments. I came across one of the structures that has a pre-built nether portal in the middle, and activated the redstone to access the two hidden chests. 
After that I decided to see where the nether portal would take me. Once on the other side a progress pop-up showed up for the casting table from one of the shards I picked up from the chests, after looking through my book I found that all info had question marks on them and that all progress was lost.

Logs:
[fml-server-1.log](http://pastebin.com/gRygVjtE)
[fml-server-2.log](http://pastebin.com/NbKHY0r5)
[fml-server-3.log](http://pastebin.com/9jpH042S)

[fml-client-latest.log](http://pastebin.com/JPThafPF)

**Comment by ReikaKalseki** [comment]
Did you lose other data (like did you get new handbooks for mods)? If so, your player NBT got partially corrupted. If not...I have no idea what would specifically target CC progress.

**Comment by FyeRee** [comment]
I did end up getting a new set of handbooks yes. Inventory was otherwise untouched.

**Comment by ReikaKalseki** [comment]
Do you have any of iChun's mods (which have been known to cause this)?

**Comment by FyeRee** [comment]
Why yes, yes I do have hats and morph. I was unaware this was an issue til now.

**Comment by TomeWyrm** [comment]
It was probably morph.

**Comment by Builderb0y** [comment]
@TomeWyrm We still have this issue on our dev server and we've uninstalled all of ichun's mods this iteration. Even if morph does cause this, it's certainly not the only cause.

**Comment by ReikaKalseki** [comment]
The thing is, our issue is presenting somewhat differently. Every instance has involved the CC dimension, not anything else. On top of that, even _that_ has stopped happening once I added debug printouts.

**Comment by HellFirePvP** [comment]
_demands reopening_

I lost my progression again today. >_>

**Comment by TomeWyrm** [comment]
Ah, this issue is why my email went nuts today.

I wish there was a bit more error correction, corruption prevention, or backup/verify type ability for player NBT. It's so damn important and so easy to permanently corrupt

**Comment by HellFirePvP** [comment]
@FyeRee Do you remember at what time this happened? The server-log is several days long, so finding sections of the log that could contain data that is relevant to this issue is not exactly easy.

**Comment by FyeRee** [comment]
@HellFirePvP ~5 minutes or less before I shutdown the server/client

**Comment by Baughn** [comment]
I had this happen, after which I got disconnections for looking at cave crystals.

The disconnections had an NBT string too long message of some kind, over 32k -- sorry, I didn't write it down at the time. Based on very little evidence, I suspect it may have been from wearing a backpack containing a bunch of books from Lost Books, each of which has a pretty massive size all by itself.

Are you serializing the entire player inventory at any point? Or, well, I'm not sure a worn backpack counts as part, but...

Maybe this helps. :X

**Comment by ReikaKalseki** [comment]
NBT does have 32KB size limits, but I have a hard time imagining they could realistically be reached...can some people check the size of their player data? I am only at 6KB, and that is for the entire thing.

**Comment by Baughn** [comment]
It's the books.

The Lost Books mod is great for immersion, but what it actually does is dump entire books from Project Gutenberg into the game. As it's a server-side mod, using the normal Minecraft book system, that creates some enormous NBTs. None of them individually break the size limit, but I had two dozen in a backpack.

The result was, as i said, a disconnection with that very specific message when i looked at cave crystals.  Probably my CC state was corrupted before then, though. I attempted to reset it, but the reset command actually gave me a null-pointer exception...

So I know what not to do now, but I have no idea what, if anything, you could do about the problem. It's an edge case anyway.

**Comment by Baughn** [comment]
The more I think about it... after disconnecting, I also ended up back at spawn, as if the game had forgotten where I was. It probably did. Hmm...

Looking at my playerdata from just before I deleted it, it's 116 kB. 45kB after compression.

I've copied it to https://madoka.brage.info/bug-reports/broken-playerdata-for-cc.dat, you can have a look if you want. It's at 5.4kB now, after the reset, which is an altogether healthier size.

**Comment by ReikaKalseki** [comment]
We do not have the mod on our server, yet still have occasional progression issues. I suspect something else is involved.

**Comment by Aivech** [comment]
Carrying around large AE2 storage cells full of items or similar nested inventories could do this.

**Comment by TomeWyrm** [comment]
How much data does the CC lexicon item take up?

Also, the question of "Are you serializing the entire player inventory at any point?" was a relatively astute one. Are you, at any point that you know of, storing something (or possibly even just reading something) in NBT that might cause you to go over the limit? Fetching the player inventory could manage that with quite a few things especially mods, but it's surely not the only way you could get large lumps of data.

**Comment by ReikaKalseki** [comment]
The CC lexicon is not very large, being at most 200 or so string tags.

**Comment by AshIndigo** [comment]
I also just got this upon reloading the world in the CC dimension. If you need info let me know.

**Comment by HellFirePvP** [comment]
Providing additional information is not a bad thing - so what did you do before logging out and back in into the CC dimension? Any information about where you were (any 'special' location), what items you had, abilities and possibly abilities that you were actively using at the time when relogging, were there any other players online? any other players in the CC dimension? [...]

**Comment by AshIndigo** [comment]
Alright so before I had logged I had been wandering around exploring and stuff. I had just recently exited one of the void rift holes (Karmir if memory serves). I had only one ability which is Lumen Dash with full lumen capacity. The major parts of my inventory were 5 practically full golden bags of holding with various loot items I have gotten and my lexicon which only has 2 notes. I play in single player mode so no one else was in the dimension except me. 

If it matters at all I had a biome ID conflict with GalaxySpace which caused the default "Luminescent Sanctuary" to appear as "Io" But I have since fixed it to have the right biome

**Comment by TomeWyrm** [comment]
Does everyone that's been getting the resets have inventory extenders (bags of holding, backpacks, etc)? Because that sounds like a clue to me ;-)

**Comment by AshIndigo** [comment]
I have 5 on me. (Little absurd I know)

**Comment by Baughn** [comment]
Iron Backpacks. Which appears to sport buggy contents-saving code.

**Comment by ReikaKalseki** [comment]
The 'backpacks' mod does not have this issue, as it in fact stores the contents in a separate file, and the item NBT only has the file ID.

**Comment by AshIndigo** [comment]
So this happened again with a different set of circumstances. I went AFK while leaving my game un-paused when I came back I had discovered my game crashed with this. http://pastebin.com/Wme45jmA
I reloaded my game and have lost my progress again. My player data NBT size is 7kb and I have 7 golden bags now. 

PS. When I loaded back up I had the falling from the sky spawn just like when you use the portal rift to enter.

**Comment by ReikaKalseki** [comment]
The crash is in the music system, though why it happened is beyond me. And we still are not much closer to finding the reason for the progression loss...

**Comment by AshIndigo** [comment]
I did it again... This time I was standing next to a Kurauri pylon when I came back I lost my progress

**Comment by AshIndigo** [comment]
My progress was wiped again after I timed out in a structure field.

**Comment by DevonX** [comment]
I just got my progression wiped when afking on the chroma collector. ._. is there any way to revert it via commands? i got 212 of 223 of the pages. Any information i can contribute to i will.

**Comment by DevonX** [comment]
This crash happned just before i lost all the progression http://pastebin.com/ywjdttFp

**Comment by Baughn** [comment]
Can't help you with reverting as-is, but you should have backups, and if you do then you can copy over the player data from the backup without altering the world.

If you don't have backups, install [AromaBackup](https://minecraft.curseforge.com/projects/aromabackup). You should have backups.

**Comment by DevonX** [comment]
Well I didn't loose my fragments so it is just to trigger the events again. I know that there is a console command to unlock everything but I want to progress normally. I am running my pack on a server (to reduce the load (I'm the only one playing on the server)) and it has only 4 gb of storage. I can manually backup the file by downloading it and storing it on my computer. I will do this in the future.

**Comment by DevonX** [comment]
It happened again. this time i got timed out in the chroma dimension when i lost it all. i managed to get all the progression back and complete the story-line normally. There is still something wrong that even tho i have made it to the end of the story i still am missing fragments that i cant unlock . i then ran the "/chromaprog debug all command" with this result (note that this is an approximate completion list that i managed to get back via commands. some triggers may remain un-triggered ) "http://i.imgur.com/pOx6WLA.png" this is when i noticed that my research level was only "basiccraft". When i tried to increase this by using the "/chromaprog level ctm true" command i got this error: "http://pastebin.com/PadCG8vR" If im not mistaken that looks like a conflict between fastcraft and the rendring of chromaticraft. I hope this info helps. I still got the list of commands that i used if you need it.

**Comment by HellFirePvP** [comment]
> i still am missing fragments that i cant unlock

Means you were, at the time, missing some progression to get those, since they're bound to progression level, not like Thaumcraft's research that sometimes requires you to scan something.

However, @ReikaKalseki i noticed something here:
https://github.com/ReikaKalseki/ChromatiCraft/blob/master/Registry/ChromaResearchManager.java#L508
Any specific reason to remove the SideOnly annotation? The interface still has it anyway.

Edit:
Additionally, it seems like the potential server-side access to the RenderItem was there since a few updates after you added the CTM progression level. Specifically it was changed from getIcon to renderIcon in this commit
https://github.com/ReikaKalseki/ChromatiCraft/commit/c29df82313db3bf496fabf26a0ccb42532ffd011#diff-bbd32f986d04d810d2025a701723e64aL503
And remained unchanged since then. While getIcon doesn't represent an issue on server side necessarily, renderIcon certainly does..

**Comment by ReikaKalseki** [comment]
Possibly fixed with a new backup system.

**Comment by lAimBotl** [comment]
Welp just lost prog.
Got this crash : https://pastebin.com/mrA4zX5c (unsure if I lost it then since I didn't focus on that, but then if i did would I have still been able to ping the repeaters with no prog? if not then I didn't lose it then.)

First time I noticed it was after I went into the aroma mining dimension,
I had a rftools builder quarrying in there, and it was chunk loaded by a chickenchunks chunk loader.
I got all the starter books, inv is the same, I also kept my old book with 184 frags.

Mods that may potentially mess with playerdata?

exu
Baubles 
ET Futurum
lycanites mobs
galacticraft
galaxyspace
hungeroverhall
multimine

Dragonapi-v22d
CRC-v22d

(also noticed all the crc entries are gone in the player dat when looking at it with a nbt editor, hmm I assume I cannot just "add" them back manually there  . . .)
